packages/mui-material/src/SvgIcon/SvgIcon.js

```diff
@@ -34,32 +35,60 @@ const SvgIconRoot = styled('svg', {
       styles[`fontSize${capitalize(ownerState.fontSize)}`],
     ];
   },
-})(({ theme, ownerState }) => ({
+})(({ theme }) => ({
   userSelect: 'none',
   width: '1em',
   height: '1em',
   display: 'inline-block',
-  // the <svg> will define the property that has `currentColor`
-  // for example heroicons uses fill="none" and stroke="currentColor"
-  fill: ownerState.hasSvgAsChild ? undefined : 'currentColor',
   flexShrink: 0,
   transition: theme.transitions?.create?.('fill', {
     duration: theme.transitions?.duration?.shorter,
   }),
-  fontSize: {
-    inherit: 'inherit',
-    small: theme.typography?.pxToRem?.(20) || '1.25rem',
-    medium: theme.typography?.pxToRem?.(24) || '1.5rem',
-    large: theme.typography?.pxToRem?.(35) || '2.1875rem',
-  }[ownerState.fontSize],
-  // TODO v5 deprecate, v6 remove for sx
-  color:
-    (theme.vars || theme).palette?.[ownerState.color]?.main ??
+  variants: [
+    {
+      props: (props) => !props.hasSvgAsChild,
+      style: {
+        // the <svg> will define the property that has `currentColor`
+        // for example heroicons uses fill="none" and stroke="currentColor"
+        fill: 'currentColor',
+      },
+    },
+    {
+      props: { fontSize: 'inherit' },
+      style: { fontSize: 'inherit' },
+    },
+    {
+      props: { fontSize: 'small' },
+      style: { fontSize: theme.typography?.pxToRem?.(20) || '1.25rem' },
+    },
+    {
+      props: { fontSize: 'medium' },
+      style: { fontSize: theme.typography?.pxToRem?.(24) || '1.5rem' },
+    },
+    {
+      props: { fontSize: 'large' },
+      style: { fontSize: theme.typography?.pxToRem?.(35) || '2.1875rem' },
+    },
+    // TODO v5 deprecate color prop, v6 remove for sx
+    ...Object.entries(theme.palette)
+      .filter(([, value]) => value.main || value.action || value.disabled)
+      .map(([color]) => ({
+        props: { color },
+        style: { color: theme.palette?.[color]?.main },
```

Comment:
```suggestion
        style: { color: (theme.vars ?? theme).palette?.[color]?.main },
```
x the other usages.

Reply:
Updated (except function calls like `theme.typography?.pxToRem()` or `theme.transitions?.create()` which don't exist in the `vars` object).